contracts/src/WarpProtocolRegistry.sol

```diff
@@ -0,0 +1,195 @@
+
+/**
+ * @dev Registry entry that represents a mapping between protocolAddress and version.
+ */
+struct ProtocolRegistryEntry {
+    uint256 version;
+    address protocolAddress;
+}
+
+/**
+ * @dev Implementation of an abstract `WarpProtocolRegistry` contract.
+ *
+ * This implementation is a contract that can be used as a base contract for protocols that are
+ * build on top of warp. It allows the protocol to be upgraded through a warp out-of-band message.
+ */
+abstract contract WarpProtocolRegistry {
+    // Address that the out-of-band warp message sets as the "source" address.
+    // The address is obviously not owned by any EOA or smart contract account, so it
+    // can not possibly be the source address of any other warp message emitted by the VM.
+    address public constant VALIDATORS_SOURCE_ADDRESS = address(0);
+
+    WarpMessenger public constant WARP_MESSENGER =
+        WarpMessenger(0x0200000000000000000000000000000000000005);
+
+    bytes32 internal immutable _chainID;
+
+    // The latest protocol version. 0 means no protocol version has been added, and isn't a valid version.
+    uint256 internal _latestVersion;
+
+    mapping(uint256 => address) internal _versionToAddress;
+    mapping(address => uint256) internal _addressToVersion;
+
+    /**
+     * @dev Emitted when a new protocol version is added to the registry.
+     */
+    event AddProtocolVersion(
+        uint256 indexed version,
+        address indexed protocolAddress
+    );
+
+    // Errors
+    error InvalidWarpMessage();
+    error InvalidSourceChainID();
+    error InvalidOriginSenderAddress();
+    error InvalidDestinationChainID();
+    error InvalidDestinationAddress();
+    error InvalidProtocolVersion();
+    error InvalidProtocolAddress();
+
+    /**
+     * @dev Initializes the contract by setting a `chainID` and `latestVersion`.
+     */
+    constructor(ProtocolRegistryEntry[] memory initialEntries) {
+        _latestVersion = 0;
+        _chainID = WARP_MESSENGER.getBlockchainID();
+
+        for (uint256 i = 0; i < initialEntries.length; i++) {
+            _addToRegistry(initialEntries[i]);
+        }
+    }
+
+    /**
+     * @dev Gets and verifies a warp out-of-band message, and adds the new protocol version
+     * address to the registry.
+     *
+     * Emits a {AddProtocolVersion} event when successful.
+     * Requirements:
+     *
+     * - a valid warp out-of-band message must be provided.
+     * - the version must be the increment of the latest version.
+     * - the protocol address must be a contract address.
+     */
+    function addProtocolVersion(uint32 messageIndex) external {
+        _addProtocolVersion(messageIndex);
+    }
+
+    /**
+     * @dev Gets the address of a protocol version.
+     * Requirements:
+     *
+     * - the version must be a valid version.
+     */
+    function getAddressFromVersion(
+        uint256 version
+    ) external view returns (address) {
+        return _getAddressFromVersion(version);
```

Comment:
What's the purpose of directly calling an internal function here and with the other external functions?

Reply:
the additional `internal` functions are to allow inherited contracts (ex. `TeleporterRegistry`) to call. If it were only `external` the contracts would have to do `this.foo()` which costs more gas.